src/main/java/randoop/generation/OrienteeringSelection.java

```diff
@@ -0,0 +1,244 @@
+ * <p>The GRT paper also does not describe how to handle input sequences that have an execution time
+ * of zero. We observe that this is the case for simple input sequences such as one that only
+ * includes the assignment of a primitive type {@code byte byte0 = (byte)1;}. We handle this by
+ * assigning these input sequences an execution time of 1 nanosecond to prevent division by zero
+ * when later computing weights.
+ */
+public class OrienteeringSelection implements InputSequenceSelector {
+  /** Map from a sequence to its weight. */
+  private final Map<Sequence, Double> weightMap = new HashMap<>();
+
+  /** Map from a sequence to the number of times it was selected. */
+  private final Map<Sequence, Integer> sequenceSelectionCount = new HashMap<>();
+
+  /**
+   * Cache for the square root of the number of method calls in a sequence of statements. Once
+   * computed for a given sequence, the value is never updated.
+   */
+  private final Map<Sequence, Double> sequenceMethodSizeSqrt = new HashMap<>();
+
+  /**
+   * Map from a sequence to its approximate execution time in nanoseconds. Once computed for a given
+   * sequence, the execution time of an input sequence is never updated.
+   */
+  private final Map<Sequence, Long> sequenceExecutionTime = new HashMap<>();
+
+  /**
+   * Bias input selection towards lower cost sequences. We first compute and update the weights of
+   * all the candidates within the candidate list before making our selection.
+   *
+   * @param candidates sequences to choose from
+   * @return the chosen sequence
+   */
+  @Override
+  public Sequence selectInputSequence(SimpleList<Sequence> candidates) {
+    double totalWeight = updateWeightMapForCandidates(candidates);
+
+    Sequence selectedSequence = Randomness.randomMemberWeighted(candidates, weightMap, totalWeight);
+    incrementCountInMap(sequenceSelectionCount, selectedSequence);
+    return selectedSequence;
+  }
+
+  /**
+   * Update the weights of the candidates in the given list. If an input sequence has not been
+   * selected or executed before, it will be assigned its default weight, computed by {@code
+   * Sequence.getSize()}. The formula for updating a sequence's weight if selection count and
+   * execution time information are both available is
+   *
+   * <p>1.0 / (k * seq.exec_time * sqrt(seq.meth_size))
+   *
+   * <p>Where k is the number of selections of seq and exec_time is the execution time of seq and
+   * meth_size is the number of method call statements in seq. This formula is a slight
+   * simplification of the one described in the GRT paper which maintains a separate exec_time for
+   * each execution of seq. However, we assume that the execution times of a sequence are the same
+   * as the first execution.
+   *
+   * @param candidates list of candidate sequences
+   * @return the total weight of all the elements in the candidate list
+   */
+  private double updateWeightMapForCandidates(SimpleList<Sequence> candidates) {
```

Comment:
Why is this called `update...` when it actually recomputes from scratch.
I think it recomputes from scratch -- is that correct?  If so, is a full recomputation from scratch necessary?

Reply:
You're right, it does recompute the weight for every candidate sequence. Initially I was solely relying on this method to do all the weight updating. However, it looks like we can take a similar approach before and selectively update weights for sequences that have been chosen and executed - and that the execution times are provided. I have made the changes to do so.